// @applitools/eyes-api/dist/input/PropertyData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropertyDataData = void 0;
const utils = require("@applitools/utils");
class PropertyDataData {
    constructor(propertyOrName, value) {
        if (utils.types.isString(propertyOrName)) {
            utils.guard.isString(propertyOrName, { name: 'name' });
            utils.guard.notNull(value, { name: 'value' });
            this._property = { name: propertyOrName, value };
        }
        else {
            utils.guard.isString(propertyOrName.name, { name: 'property.name' });
            utils.guard.notNull(propertyOrName.value, { name: 'property.value' });
            this._property = propertyOrName;
        }
    }
    get name() {
        return this._property.name;
    }
    set name(name) {
        this._property.name = name;
    }
    getName() {
        return this.name;
    }
    setName(name) {
        this.name = name;
    }
    get value() {
        return this._property.value;
    }
    set value(value) {
        this._property.value = value;
    }
    getValue() {
        return this.value;
    }
    setValue(value) {
        this.value = value;
    }
    /** @internal */
    toObject() {
        return this._property;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._property);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.PropertyDataData = PropertyDataData;
