// @applitools/eyes-api/dist/src/BatchClose.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchClose = exports.closeBatch = void 0;
const utils = require("@applitools/utils");
function closeBatch(spec) {
    return (options) => {
        utils.guard.notNull(options.batchIds, { name: 'options.batchIds' });
        return spec.closeBatches({
            batchIds: options.batchIds,
            serverUrl: options.serverUrl,
            apiKey: options.apiKey,
            proxy: options.proxy,
        });
    };
}
exports.closeBatch = closeBatch;
class BatchClose {
    constructor(options) {
        this._options = { batchIds: null };
        if (options)
            this._options = options;
    }
    get _spec() {
        return this.constructor._spec;
    }
    static async close(options) {
        utils.guard.notNull(options.batchIds, { name: 'options.batchIds' });
        await this._spec.closeBatches(options);
    }
    async close() {
        utils.guard.notNull(this._options.batchIds, { name: 'batchIds' });
        await this._spec.closeBatches(this._options);
    }
    setBatchIds(batchIds) {
        this._options.batchIds = batchIds;
        return this;
    }
    setUrl(serverUrl) {
        this._options.serverUrl = serverUrl;
        return this;
    }
    setApiKey(apiKey) {
        this._options.apiKey = apiKey;
        return this;
    }
    setProxy(proxy) {
        this._options.proxy = proxy;
        return this;
    }
}
exports.BatchClose = BatchClose;
