// @applitools/eyes-api/dist/src/output/MatchResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MatchResultData = void 0;
const utils = require("@applitools/utils");
class MatchResultData {
    /** @internal */
    constructor(result) {
        this._result = {};
        if (!result)
            return this;
        this._result = result instanceof MatchResultData ? result.toJSON() : result;
    }
    get asExpected() {
        return this._result.asExpected;
    }
    getAsExpected() {
        return this.asExpected;
    }
    /** @deprecated */
    setAsExpected(asExpected) {
        this._result.asExpected = asExpected;
    }
    get windowId() {
        return this._result.windowId;
    }
    getWindowId() {
        return this.windowId;
    }
    /** @deprecated */
    setWindowId(windowId) {
        this._result.windowId = windowId;
    }
    /** @internal */
    toObject() {
        return this._result;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._result);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.MatchResultData = MatchResultData;
