// @applitools/eyes-api/dist/src/output/SessionUrls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionUrlsData = void 0;
const utils = require("@applitools/utils");
class SessionUrlsData {
    /** @internal */
    constructor(urls) {
        this._urls = {};
        if (!urls)
            return this;
        this._urls = urls instanceof SessionUrlsData ? urls.toJSON() : urls;
    }
    get batch() {
        return this._urls.batch;
    }
    getBatch() {
        return this.batch;
    }
    /** @deprecated */
    setBatch(batch) {
        this._urls.batch = batch;
    }
    get session() {
        return this._urls.session;
    }
    getSession() {
        return this.session;
    }
    /** @deprecated */
    setSession(session) {
        this._urls.session = session;
    }
    /** @internal */
    toObject() {
        return this._urls;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._urls);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.SessionUrlsData = SessionUrlsData;
