// @applitools/eyes-api/dist/src/output/StepInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepInfoData = void 0;
const utils = require("@applitools/utils");
const AppUrls_1 = require("./AppUrls");
const ApiUrls_1 = require("./ApiUrls");
class StepInfoData {
    /** @internal */
    constructor(info) {
        this._info = {};
        if (!info)
            return this;
        this._info = info instanceof StepInfoData ? info.toJSON() : info;
    }
    get name() {
        return this._info.name;
    }
    getName() {
        return this.name;
    }
    /** @deprecated */
    setName(value) {
        this._info.name = value;
    }
    get isDifferent() {
        return this._info.isDifferent;
    }
    getIsDifferent() {
        return this.isDifferent;
    }
    /** @deprecated */
    setIsDifferent(value) {
        this._info.isDifferent = value;
    }
    get hasBaselineImage() {
        return this._info.hasBaselineImage;
    }
    getHasBaselineImage() {
        return this.hasBaselineImage;
    }
    /** @deprecated */
    setHasBaselineImage(value) {
        this._info.hasBaselineImage = value;
    }
    get hasCurrentImage() {
        return this._info.hasCurrentImage;
    }
    getHasCurrentImage() {
        return this.hasCurrentImage;
    }
    /** @deprecated */
    setHasCurrentImage(hasCurrentImage) {
        this._info.hasCurrentImage = hasCurrentImage;
    }
    get appUrls() {
        return this._info.appUrls;
    }
    getAppUrls() {
        return new AppUrls_1.AppUrlsData(this.appUrls);
    }
    /** @deprecated */
    setAppUrls(appUrls) {
        this._info.appUrls = appUrls;
    }
    get apiUrls() {
        return this._info.apiUrls;
    }
    getApiUrls() {
        return new ApiUrls_1.ApiUrlsData(this.apiUrls);
    }
    /** @deprecated */
    setApiUrls(apiUrls) {
        this._info.apiUrls = apiUrls;
    }
    get renderId() {
        return this._info.renderId;
    }
    getRenderId() {
        return this.renderId;
    }
    /** @deprecated */
    setRenderId(renderId) {
        this._info.renderId = renderId;
    }
    /** @internal */
    toObject() {
        return this._info;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._info);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.StepInfoData = StepInfoData;
