// @wdio/applitools-service/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("@wdio/logger"));
const eyes_webdriverio_1 = require("@applitools/eyes-webdriverio");
const log = logger_1.default('@wdio/applitools-service');
const DEFAULT_VIEWPORT = {
    width: 1440,
    height: 900
};
class ApplitoolsService {
    constructor(_options) {
        this._options = _options;
        this._isConfigured = false;
        this._eyes = new eyes_webdriverio_1.Eyes();
    }
    /**
     * set API key in onPrepare hook and start test
     */
    beforeSession() {
        const key = this._options.key || process.env.APPLITOOLS_KEY;
        const serverUrl = this._options.serverUrl || process.env.APPLITOOLS_SERVER_URL;
        if (!key) {
            throw new Error('Couldn\'t find an Applitools "applitools.key" in config nor "APPLITOOLS_KEY" in the environment');
        }
        /**
         * Optionally set a specific server url
         */
        if (serverUrl) {
            this._eyes.setServerUrl(serverUrl);
        }
        this._isConfigured = true;
        this._eyes.setApiKey(key);
        if (this._options.eyesProxy) {
            this._eyes.setProxy(this._options.eyesProxy);
        }
        this._viewport = Object.assign({ ...DEFAULT_VIEWPORT }, this._options.viewport);
    }
    /**
     * set custom commands
     */
    before(caps, specs, browser) {
        this._browser = browser;
        if (!this._isConfigured) {
            return;
        }
        this._browser.addCommand('takeSnapshot', this._takeSnapshot.bind(this));
        this._browser.addCommand('takeRegionSnapshot', this._takeRegionSnapshot.bind(this));
    }
    _takeSnapshot(title) {
        if (!title) {
            throw new Error('A title for the Applitools snapshot is missing');
        }
        return this._eyes.check(title, eyes_webdriverio_1.Target.window());
    }
    _takeRegionSnapshot(title, region, frame) {
        if (!title) {
            throw new Error('A title for the Applitools snapshot is missing');
        }
        if (!region) {
            throw new Error('A region for the Applitools snapshot is missing');
        }
        if (!frame) {
            return this._eyes.check(title, eyes_webdriverio_1.Target.region(region));
        }
        return this._eyes.check(title, eyes_webdriverio_1.Target.region(region, frame));
    }
    beforeTest(test) {
        if (!this._isConfigured || !this._browser) {
            return;
        }
        log.info(`Open eyes for ${test.parent} ${test.title}`);
        this._browser.call(() => this._eyes.open(this._browser, test.title, test.parent, this._viewport));
    }
    afterTest() {
        if (!this._isConfigured || !this._browser) {
            return;
        }
        this._browser.call(this._eyes.close.bind(this._eyes));
    }
    after() {
        if (!this._isConfigured || !this._browser) {
            return;
        }
        this._browser.call(this._eyes.abortIfNotClosed.bind(this._eyes));
    }
}
exports.default = ApplitoolsService;
__exportStar(require("./types"), exports);
