// @applitools/eyes-api/dist/input/CheckSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckSettingsFluent = void 0;
const utils = require("@applitools/utils");
const AccessibilityRegionType_1 = require("../enums/AccessibilityRegionType");
const MatchLevel_1 = require("../enums/MatchLevel");
class CheckSettingsFluent {
    constructor(settings) {
        this._settings = {};
        if (!settings)
            return this;
        if (settings.name)
            this.name(settings.name);
        if (settings.region)
            this.region(settings.region);
        if (settings.frames) {
            settings.frames.forEach(reference => {
                if (utils.types.isNull(reference))
                    return;
                if (utils.types.has(reference, 'frame')) {
                    this.frame(reference.frame, reference.scrollRootElement);
                }
                else {
                    this.frame(reference);
                }
            });
        }
        if (settings.scrollRootElement)
            this.scrollRootElement(settings.scrollRootElement);
        if (!utils.types.isNull(settings.fully))
            this.fully(settings.fully);
        if (settings.matchLevel)
            this.matchLevel(settings.matchLevel);
        if (!utils.types.isNull(settings.useDom))
            this.useDom(settings.useDom);
        if (!utils.types.isNull(settings.sendDom))
            this.sendDom(settings.sendDom);
        if (!utils.types.isNull(settings.enablePatterns))
            this.enablePatterns(settings.enablePatterns);
        if (!utils.types.isNull(settings.ignoreDisplacements))
            this.ignoreDisplacements(settings.ignoreDisplacements);
        if (!utils.types.isNull(settings.ignoreCaret))
            this.ignoreCaret(settings.ignoreCaret);
        if (settings.ignoreRegions) {
            settings.ignoreRegions.forEach(ignoreRegion => this.ignoreRegion(ignoreRegion));
        }
        if (settings.layoutRegions) {
            settings.layoutRegions.forEach(layoutRegion => this.layoutRegion(layoutRegion));
        }
        if (settings.strictRegions) {
            settings.strictRegions.forEach(strictRegion => this.strictRegion(strictRegion));
        }
        if (settings.contentRegions) {
            settings.contentRegions.forEach(contentRegion => this.contentRegion(contentRegion));
        }
        if (settings.floatingRegions) {
            settings.floatingRegions.forEach(floatingRegion => this.floatingRegion(floatingRegion));
        }
        if (settings.accessibilityRegions) {
            settings.accessibilityRegions.forEach(accessibilityRegion => this.accessibilityRegion(accessibilityRegion));
        }
        if (!utils.types.isNull(settings.disableBrowserFetching))
            this.disableBrowserFetching(settings.disableBrowserFetching);
        if (!utils.types.isNull(settings.layoutBreakpoints))
            this.layoutBreakpoints(settings.layoutBreakpoints);
        if (settings.hooks) {
            Object.entries(settings.hooks).forEach(([name, script]) => this.hook(name, script));
        }
        if (settings.visualGridOptions) {
            Object.entries(settings.visualGridOptions).forEach(([key, value]) => this.visualGridOption(key, value));
        }
        if (settings.renderId)
            this.renderId(settings.renderId);
        if (settings.variationGroupId)
            this.variationGroupId(settings.variationGroupId);
        if (!utils.types.isNull(settings.timeout))
            this.timeout(settings.timeout);
    }
    /** @internal */
    static window() {
        return new this();
    }
    /** @internal */
    static region(region) {
        return new this().region(region);
    }
    /** @internal */
    static frame(contextOrFrame, scrollRootElement) {
        return new this().frame(contextOrFrame, scrollRootElement);
    }
    get _spec() {
        return this.constructor._spec;
    }
    _isFrameReference(value) {
        return utils.types.isNumber(value) || utils.types.isString(value) || this._isElementReference(value);
    }
    _isRegionReference(value) {
        return (utils.types.has(value, ['x', 'y', 'width', 'height']) ||
            utils.types.has(value, ['left', 'top', 'width', 'height']) ||
            this._isElementReference(value));
    }
    _isElementReference(value) {
        return this._spec.isElement(value) || this._spec.isSelector(value);
    }
    /** @undocumented */
    name(name) {
        utils.guard.isString(name, { name: 'name' });
        this._settings.name = name;
        return this;
    }
    withName(name) {
        return this.name(name);
    }
    region(region) {
        utils.guard.custom(region, value => this._isRegionReference(value), { name: 'region' });
        this._settings.region = region;
        return this;
    }
    frame(contextOrFrame, scrollRootElement) {
        const context = utils.types.has(contextOrFrame, 'frame')
            ? contextOrFrame
            : { frame: contextOrFrame, scrollRootElement };
        if (!this._settings.frames)
            this._settings.frames = [];
        utils.guard.custom(context.frame, value => this._isFrameReference(value), { name: 'frame' });
        utils.guard.custom(context.scrollRootElement, value => this._isElementReference(value), {
            name: 'scrollRootElement',
            strict: false,
        });
        this._settings.frames.push(context);
        return this;
    }
    ignoreRegion(ignoreRegion) {
        if (!this._settings.ignoreRegions)
            this._settings.ignoreRegions = [];
        this._settings.ignoreRegions.push(ignoreRegion);
        return this;
    }
    ignoreRegions(...ignoreRegions) {
        ignoreRegions.forEach(ignoreRegion => this.ignoreRegion(ignoreRegion));
        return this;
    }
    /** @deprecated */
    ignore(ignoreRegion) {
        return this.ignoreRegion(ignoreRegion);
    }
    /** @deprecated */
    ignores(...ignoreRegions) {
        return this.ignoreRegions(...ignoreRegions);
    }
    layoutRegion(layoutRegion) {
        if (!this._settings.layoutRegions)
            this._settings.layoutRegions = [];
        this._settings.layoutRegions.push(layoutRegion);
        return this;
    }
    layoutRegions(...layoutRegions) {
        layoutRegions.forEach(layoutRegion => this.layoutRegion(layoutRegion));
        return this;
    }
    strictRegion(strictRegion) {
        if (!this._settings.strictRegions)
            this._settings.strictRegions = [];
        this._settings.strictRegions.push(strictRegion);
        return this;
    }
    strictRegions(...regions) {
        regions.forEach(region => this.strictRegion(region));
        return this;
    }
    contentRegion(region) {
        if (!this._settings.contentRegions)
            this._settings.contentRegions = [];
        this._settings.contentRegions.push(region);
        return this;
    }
    contentRegions(...regions) {
        regions.forEach(region => this.contentRegion(region));
        return this;
    }
    floatingRegion(region, maxUpOffset, maxDownOffset, maxLeftOffset, maxRightOffset) {
        const floatingRegion = utils.types.has(region, 'region')
            ? region
            : { region, maxUpOffset, maxDownOffset, maxLeftOffset, maxRightOffset };
        utils.guard.custom(floatingRegion.region, value => this._isRegionReference(value), {
            name: 'region',
        });
        utils.guard.isNumber(floatingRegion.maxUpOffset, { name: 'maxUpOffset' });
        utils.guard.isNumber(floatingRegion.maxDownOffset, { name: 'maxDownOffset' });
        utils.guard.isNumber(floatingRegion.maxLeftOffset, { name: 'maxLeftOffset' });
        utils.guard.isNumber(floatingRegion.maxRightOffset, { name: 'maxRightOffset' });
        if (!this._settings.floatingRegions)
            this._settings.floatingRegions = [];
        this._settings.floatingRegions.push(floatingRegion);
        return this;
    }
    floatingRegions(regionOrMaxOffset, ...regions) {
        if (utils.types.isNumber(regionOrMaxOffset)) {
            const maxOffset = regionOrMaxOffset;
            regions.forEach((region) => this.floatingRegion({
                region,
                maxUpOffset: maxOffset,
                maxDownOffset: maxOffset,
                maxLeftOffset: maxOffset,
                maxRightOffset: maxOffset,
            }));
        }
        else {
            this.floatingRegion(regionOrMaxOffset);
            regions.forEach((floatingRegion) => this.floatingRegion(floatingRegion));
        }
        return this;
    }
    floating(region, maxUpOffset, maxDownOffset, maxLeftOffset, maxRightOffset) {
        return this.floatingRegion(region, maxUpOffset, maxDownOffset, maxLeftOffset, maxRightOffset);
    }
    floatings(regionOrMaxOffset, ...regions) {
        return this.floatingRegions(regionOrMaxOffset, ...regions);
    }
    accessibilityRegion(region, type) {
        const accessibilityRegion = utils.types.has(region, 'region') ? region : { region, type };
        utils.guard.custom(accessibilityRegion.region, value => this._isRegionReference(value), {
            name: 'region',
        });
        utils.guard.isEnumValue(accessibilityRegion.type, AccessibilityRegionType_1.AccessibilityRegionTypeEnum, {
            name: 'type',
            strict: false,
        });
        if (!this._settings.accessibilityRegions)
            this._settings.accessibilityRegions = [];
        this._settings.accessibilityRegions.push(accessibilityRegion);
        return this;
    }
    accessibilityRegions(regionOrType, ...regions) {
        if (utils.types.isEnumValue(regionOrType, AccessibilityRegionType_1.AccessibilityRegionTypeEnum)) {
            const type = regionOrType;
            regions.forEach((region) => this.accessibilityRegion({ region, type }));
        }
        else {
            this.accessibilityRegion(regionOrType);
            regions.forEach((floatingRegion) => this.accessibilityRegion(floatingRegion));
        }
        return this;
    }
    scrollRootElement(scrollRootElement) {
        utils.guard.custom(scrollRootElement, value => this._isElementReference(value), {
            name: 'scrollRootElement',
        });
        if (this._settings.frames && this._settings.frames.length > 0) {
            const context = this._settings.frames[this._settings.frames.length - 1];
            context.scrollRootElement = scrollRootElement;
        }
        this._settings.scrollRootElement = scrollRootElement;
        return this;
    }
    fully(fully = true) {
        utils.guard.isBoolean(fully, { name: 'fully' });
        this._settings.fully = fully;
        return this;
    }
    /** @deprecated */
    stitchContent(stitchContent = true) {
        return this.fully(stitchContent);
    }
    matchLevel(matchLevel) {
        utils.guard.isEnumValue(matchLevel, MatchLevel_1.MatchLevelEnum, { name: 'matchLevel' });
        this._settings.matchLevel = matchLevel;
        return this;
    }
    layout() {
        this._settings.matchLevel = MatchLevel_1.MatchLevelEnum.Layout;
        return this;
    }
    exact() {
        this._settings.matchLevel = MatchLevel_1.MatchLevelEnum.Exact;
        return this;
    }
    strict() {
        this._settings.matchLevel = MatchLevel_1.MatchLevelEnum.Strict;
        return this;
    }
    content() {
        this._settings.matchLevel = MatchLevel_1.MatchLevelEnum.Content;
        return this;
    }
    useDom(useDom = true) {
        utils.guard.isBoolean(useDom, { name: 'useDom' });
        this._settings.useDom = useDom;
        return this;
    }
    sendDom(sendDom = true) {
        utils.guard.isBoolean(sendDom, { name: 'sendDom' });
        this._settings.sendDom = sendDom;
        return this;
    }
    enablePatterns(enablePatterns = true) {
        utils.guard.isBoolean(enablePatterns, { name: 'enablePatterns' });
        this._settings.enablePatterns = enablePatterns;
        return this;
    }
    ignoreDisplacements(ignoreDisplacements = true) {
        utils.guard.isBoolean(ignoreDisplacements, { name: 'ignoreDisplacements' });
        this._settings.ignoreDisplacements = ignoreDisplacements;
        return this;
    }
    ignoreCaret(ignoreCaret = true) {
        utils.guard.isBoolean(ignoreCaret, { name: 'ignoreCaret' });
        this._settings.ignoreCaret = ignoreCaret;
        return this;
    }
    disableBrowserFetching(disableBrowserFetching) {
        utils.guard.isBoolean(disableBrowserFetching, { name: 'disableBrowserFetching' });
        this._settings.disableBrowserFetching = disableBrowserFetching;
        return this;
    }
    layoutBreakpoints(layoutBreakpoints = true) {
        if (!utils.types.isArray(layoutBreakpoints)) {
            this._settings.layoutBreakpoints = layoutBreakpoints;
        }
        else if (layoutBreakpoints.length === 0) {
            this._settings.layoutBreakpoints = false;
        }
        else {
            this._settings.layoutBreakpoints = Array.from(new Set(layoutBreakpoints)).sort((a, b) => (a < b ? 1 : -1));
        }
        return this;
    }
    hook(name, script) {
        this._settings.hooks = Object.assign(Object.assign({}, this._settings.hooks), { [name]: script });
        return this;
    }
    beforeRenderScreenshotHook(script) {
        return this.hook('beforeCaptureScreenshot', script);
    }
    /** @deprecated */
    webHook(script) {
        return this.beforeRenderScreenshotHook(script);
    }
    visualGridOption(key, value) {
        this._settings.visualGridOptions = Object.assign(Object.assign({}, this._settings.visualGridOptions), { [key]: value });
        return this;
    }
    visualGridOptions(options) {
        this._settings.visualGridOptions = options;
        return this;
    }
    renderId(renderId) {
        utils.guard.isString(renderId, { name: 'renderId' });
        this._settings.renderId = renderId;
        return this;
    }
    variationGroupId(variationGroupId) {
        utils.guard.isString(variationGroupId, { name: 'variationGroupId' });
        this._settings.variationGroupId = variationGroupId;
        return this;
    }
    timeout(timeout) {
        utils.guard.isNumber(timeout, { name: 'timeout' });
        this._settings.timeout = timeout;
        return this;
    }
    /** @internal */
    toObject() {
        return this._settings;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._settings);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.CheckSettingsFluent = CheckSettingsFluent;
