// @applitools/eyes-api/dist/input/Location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocationData = void 0;
const utils = require("@applitools/utils");
class LocationData {
    constructor(locationOrX, y) {
        if (utils.types.isNumber(locationOrX)) {
            return new LocationData({ x: locationOrX, y });
        }
        this._location = {};
        this.x = locationOrX.x;
        this.y = locationOrX.y;
    }
    get x() {
        return this._location.x;
    }
    set x(x) {
        utils.guard.isNumber(x, { name: 'x' });
        this._location.x = x;
    }
    getX() {
        return this.x;
    }
    setX(x) {
        this.x = x;
    }
    get y() {
        return this._location.y;
    }
    set y(y) {
        utils.guard.isNumber(y, { name: 'y' });
        this._location.y = y;
    }
    getY() {
        return this.y;
    }
    setY(y) {
        this.y = y;
    }
    /** @internal */
    toObject() {
        return this._location;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._location);
    }
    /** @internal */
    toString() {
        return `(${this.x}, ${this.y})`;
    }
}
exports.LocationData = LocationData;
