// @applitools/eyes-api/dist/src/input/ProxySettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxySettingsData = void 0;
const utils = require("@applitools/utils");
class ProxySettingsData {
    constructor(proxyOrUrl, username, password, isHttpOnly) {
        this._proxy = {};
        utils.guard.notNull(proxyOrUrl, { name: 'proxyOrUrl' });
        if (utils.types.isString(proxyOrUrl)) {
            return new ProxySettingsData({ url: proxyOrUrl, username, password, isHttpOnly });
        }
        this._proxy = proxyOrUrl;
    }
    get url() {
        return this._proxy.url;
    }
    getUri() {
        return this.url;
    }
    getUrl() {
        return this.url;
    }
    get username() {
        return this._proxy.username;
    }
    getUsername() {
        return this.username;
    }
    get password() {
        return this.password;
    }
    getPassword() {
        return this.password;
    }
    get isHttpOnly() {
        return this._proxy.isHttpOnly;
    }
    getIsHttpOnly() {
        return this.isHttpOnly;
    }
    /** @internal */
    toObject() {
        return this._proxy;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._proxy);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.ProxySettingsData = ProxySettingsData;
