// @applitools/eyes-api/dist/src/input/RectangleSize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RectangleSizeData = void 0;
const utils = require("@applitools/utils");
class RectangleSizeData {
    constructor(sizeOrWidth, height) {
        this._size = {};
        if (utils.types.isNumber(sizeOrWidth)) {
            return new RectangleSizeData({ width: sizeOrWidth, height });
        }
        this.width = sizeOrWidth.width;
        this.height = sizeOrWidth.height;
    }
    get width() {
        return this._size.width;
    }
    set width(width) {
        utils.guard.isNumber(width, { name: 'width', gte: 0 });
        this._size.width = width;
    }
    getWidth() {
        return this.width;
    }
    setWidth(width) {
        this.width = width;
    }
    get height() {
        return this._size.height;
    }
    set height(height) {
        utils.guard.isNumber(height, { name: 'height', gte: 0 });
        this._size.height = height;
    }
    getHeight() {
        return this.height;
    }
    setHeight(height) {
        this.height = height;
    }
    /** @internal */
    toObject() {
        return this._size;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._size);
    }
    /** @internal */
    toString() {
        return `${this.width}x${this.height}`;
    }
}
exports.RectangleSizeData = RectangleSizeData;
