// @applitools/eyes-api/dist/output/ApiUrls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiUrlsData = void 0;
const utils = require("@applitools/utils");
class ApiUrlsData {
    /** @internal */
    constructor(urls) {
        this._urls = {};
        if (!urls)
            return this;
        this._urls = urls instanceof ApiUrlsData ? urls.toJSON() : urls;
    }
    get baselineImage() {
        return this._urls.baselineImage;
    }
    getBaselineImage() {
        return this.baselineImage;
    }
    /** @deprecated */
    setBaselineImage(setBaselineImage) {
        this._urls.baselineImage = setBaselineImage;
    }
    get currentImage() {
        return this._urls.currentImage;
    }
    getCurrentImage() {
        return this.currentImage;
    }
    /** @deprecated */
    setCurrentImage(currentImage) {
        this._urls.currentImage = currentImage;
    }
    get checkpointImage() {
        return this._urls.checkpointImage;
    }
    getCheckpointImage() {
        return this.checkpointImage;
    }
    /** @deprecated */
    setCheckpointImage(checkpointImage) {
        this._urls.checkpointImage = checkpointImage;
    }
    get checkpointImageThumbnail() {
        return this._urls.checkpointImageThumbnail;
    }
    getCheckpointImageThumbnail() {
        return this.checkpointImageThumbnail;
    }
    /** @deprecated */
    setCheckpointImageThumbnail(checkpointImageThumbnail) {
        this._urls.checkpointImageThumbnail = checkpointImageThumbnail;
    }
    get diffImage() {
        return this._urls.diffImage;
    }
    getDiffImage() {
        return this.diffImage;
    }
    /** @deprecated */
    setDiffImage(diffImage) {
        this._urls.diffImage = diffImage;
    }
    /** @internal */
    toObject() {
        return this._urls;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._urls);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.ApiUrlsData = ApiUrlsData;
