// @applitools/eyes-api/dist/src/enums/BrowserType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserType = void 0;
var BrowserType;
(function (BrowserType) {
    BrowserType["CHROME"] = "chrome";
    BrowserType["CHROME_ONE_VERSION_BACK"] = "chrome-one-version-back";
    BrowserType["CHROME_TWO_VERSIONS_BACK"] = "chrome-two-versions-back";
    BrowserType["FIREFOX"] = "firefox";
    BrowserType["FIREFOX_ONE_VERSION_BACK"] = "firefox-one-version-back";
    BrowserType["FIREFOX_TWO_VERSIONS_BACK"] = "firefox-two-versions-back";
    BrowserType["IE_11"] = "ie";
    BrowserType["IE_10"] = "ie10";
    BrowserType["EDGE"] = "edge";
    BrowserType["EDGE_CHROMIUM"] = "edgechromium";
    BrowserType["EDGE_LEGACY"] = "edgelegacy";
    BrowserType["EDGE_CHROMIUM_ONE_VERSION_BACK"] = "edgechromium-one-version-back";
    BrowserType["EDGE_CHROMIUM_TWO_VERSIONS_BACK"] = "edgechromium-two-versions-back";
    BrowserType["SAFARI"] = "safari";
    BrowserType["SAFARI_EARLY_ACCESS"] = "safari-earlyaccess";
    BrowserType["SAFARI_ONE_VERSION_BACK"] = "safari-one-version-back";
    BrowserType["SAFARI_TWO_VERSIONS_BACK"] = "safari-two-versions-back";
})(BrowserType = exports.BrowserType || (exports.BrowserType = {}));
