// @applitools/eyes-api/dist/output/AppUrls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppUrlsData = void 0;
const utils = require("@applitools/utils");
class AppUrlsData {
    /** @internal */
    constructor(urls) {
        this._urls = {};
        if (!urls)
            return this;
        this._urls = urls instanceof AppUrlsData ? urls.toJSON() : urls;
    }
    get step() {
        return this._urls.step;
    }
    getStep() {
        return this.step;
    }
    /** @deprecated */
    setStep(step) {
        this._urls.step = step;
    }
    get stepEditor() {
        return this._urls.stepEditor;
    }
    getStepEditor() {
        return this.stepEditor;
    }
    /** @deprecated */
    setStepEditor(stepEditor) {
        this._urls.stepEditor = stepEditor;
    }
    /** @internal */
    toObject() {
        return this._urls;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._urls);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.AppUrlsData = AppUrlsData;
