// @applitools/eyes-api/dist/output/RenderingInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderingInfoData = void 0;
const utils = require("@applitools/utils");
class RenderingInfoData {
    /** @internal */
    constructor(info) {
        this._info = {};
        this._info = info instanceof RenderingInfoData ? info.toJSON() : info;
    }
    get accessToken() {
        return this._info.accessToken;
    }
    getAccessToken() {
        return this.accessToken;
    }
    getDecodedAccessToken() {
        return utils.general.jwtDecode(this._info.accessToken);
    }
    /** @deprecated */
    setAccessToken(accessToken) {
        this._info.accessToken = accessToken;
    }
    get serviceUrl() {
        return this._info.serviceUrl;
    }
    getServiceUrl() {
        return this.serviceUrl;
    }
    /** @deprecated */
    setServiceUrl(serviceUrl) {
        this._info.serviceUrl = serviceUrl;
    }
    get resultsUrl() {
        return this._info.resultsUrl;
    }
    getResultsUrl() {
        return this.resultsUrl;
    }
    /** @deprecated */
    setResultsUrl(resultsUrl) {
        this._info.resultsUrl = resultsUrl;
    }
    get stitchingServiceUrl() {
        return this._info.stitchingServiceUrl;
    }
    getStitchingServiceUrl() {
        return this.stitchingServiceUrl;
    }
    /** @deprecated */
    setStitchingServiceUrl(stitchingServiceUrl) {
        this._info.stitchingServiceUrl = stitchingServiceUrl;
    }
    /** @internal */
    toObject() {
        return this._info;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._info);
    }
    /** @internal */
    toString() {
        return utils.general.toString(this);
    }
}
exports.RenderingInfoData = RenderingInfoData;
