// @applitools/eyes-api/dist/output/TestResultContainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestResultContainerData = void 0;
const utils = require("@applitools/utils");
const TestResults_1 = require("./TestResults");
class TestResultContainerData {
    /** @internal */
    constructor(container) {
        this._container = {};
        this._container = container instanceof TestResultContainerData ? container.toJSON() : container;
    }
    get testResults() {
        return this._container.testResults;
    }
    getTestResults() {
        return new TestResults_1.TestResultsData(this.testResults);
    }
    get exception() {
        return this._container.exception;
    }
    getException() {
        return this.exception;
    }
    /** @internal */
    toObject() {
        return this._container;
    }
    /** @internal */
    toJSON() {
        return utils.general.toJSON(this._container);
    }
    /** @internal */
    toString() {
        return `${this.testResults ? this.testResults : ''} - ${this.exception ? this.exception : ''}`;
    }
}
exports.TestResultContainerData = TestResultContainerData;
