// @applitools/eyes-selenium/dist/api.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchClose = exports.Target = exports.CheckSettings = exports.Configuration = exports.Eyes = void 0;
const eyes_sdk_core_1 = require("@applitools/eyes-sdk-core");
const api = require("@applitools/eyes-api");
const spec = require("./spec-driver");
const sdk = eyes_sdk_core_1.makeSDK({
    name: 'eyes.selenium',
    version: `${require('../package.json').version}--${process.env.APPLITOOLS_SELENIUM_MAJOR_VERSION}`,
    spec,
    VisualGridClient: require('@applitools/visual-grid-client'),
});
__exportStar(require("@applitools/eyes-api"), exports);
class Eyes extends api.Eyes {
}
exports.Eyes = Eyes;
Eyes._spec = sdk;
class Configuration extends api.Configuration {
}
exports.Configuration = Configuration;
Configuration._spec = sdk;
class CheckSettings extends api.CheckSettings {
}
exports.CheckSettings = CheckSettings;
CheckSettings._spec = sdk;
exports.Target = CheckSettings;
class BatchClose extends api.BatchClose {
}
exports.BatchClose = BatchClose;
BatchClose._spec = sdk;
